omniledger/service/viewchange.go

```diff
@@ -0,0 +1,371 @@
+	sendNV viewchange.SendNewViewReqFunc, isLeader viewchange.IsLeaderFunc, k string) {
+	m.Lock()
+	defer m.Unlock()
+	c := viewchange.NewController(sendVC, sendNV, isLeader)
+	m.controllers[k] = &c
+}
+
+func (m *viewChangeManager) start(myID network.ServerIdentityID, scID skipchain.SkipBlockID, initialDuration time.Duration, f int, k string) {
+	m.Lock()
+	defer m.Unlock()
+	c := m.controllers[k]
+	go c.Start(myID, scID, initialDuration, f)
+}
+
+func (m *viewChangeManager) addReq(req viewchange.InitReq) {
+	m.Lock()
+	defer m.Unlock()
+	c := m.controllers[string(req.View.Gen)]
+	c.AddReq(req)
+}
+
+func (m *viewChangeManager) addAnomaly(req viewchange.InitReq) {
+	m.Lock()
+	defer m.Unlock()
+	c := m.controllers[string(req.View.Gen)]
+	c.AddAnomaly(req)
+}
+
+func (m *viewChangeManager) done(view viewchange.View) {
+	m.Lock()
+	defer m.Unlock()
+	c := m.controllers[string(view.Gen)]
+	c.Done(view)
+	log.Lvl3("view-change done for " + view.String())
+}
+
+func (m *viewChangeManager) waiting(k string) bool {
+	m.Lock()
+	defer m.Unlock()
+	c := m.controllers[k]
+	return c.Waiting()
+}
+
+func (m *viewChangeManager) closeAll() {
+	m.Lock()
+	defer m.Unlock()
+	for _, c := range m.controllers {
+		c.Stop()
+	}
+	m.controllers = make(map[string]*viewchange.Controller)
+}
+
+// sendViewChangeReq is called when the node detects that a view change is
+// needed. It uses SendRaw to send the message to all other nodes. This
+// function should only be used as a callback in ViewChangeLog.
+func (s *Service) sendViewChangeReq(view viewchange.View) error {
+	log.Lvl2(s.ServerIdentity(), "sending view-change request for view:", view)
+	latest, err := s.db().GetLatestByID(view.ID)
+	if err != nil {
+		return err
+	}
+	req := viewchange.InitReq{
+		SignerID: s.ServerIdentity().ID,
+		View:     view,
+	}
+	if err := req.Sign(s.getPrivateKey()); err != nil {
+		return err
+	}
+	for _, sid := range latest.Roster.List {
+		if sid.Equal(s.ServerIdentity()) {
+			continue
+		}
+		if err := s.SendRaw(sid, &req); err != nil {
+			// Having an error here is fine because not all the
+			// nodes are guaranteed to be online. So we log a
+			// warning instead of returning an error.
+			log.Warn(s.ServerIdentity(), err)
+		}
+	}
+	return nil
+}
+
+func (s *Service) sendNewView(proof []viewchange.InitReq) {
+
+	if len(proof) == 0 {
+		log.Error(s.ServerIdentity(), "not enough proofs")
+	}
+	log.Lvl2(s.ServerIdentity(), "sending new-view request for view:", proof[0].View)
+
+	// Our own proof might not be signed, so sign it.
+	for i := range proof {
+		if proof[i].SignerID.Equal(s.ServerIdentity().ID) && len(proof[i].Signature) == 0 {
+			proof[i].Sign(s.getPrivateKey())
+		}
+	}
+
+	sb := s.db().GetByID(proof[0].View.ID)
+	req := viewchange.NewViewReq{
+		Roster: *rotateRoster(sb.Roster, proof[0].View.LeaderIndex),
+		Proof:  proof,
+	}
+
+	go func() {
+		// This go-routine eventually exists because both cosi and
```

Comment:
s/exists/exits/

Uncle Dave Cheny would be happy. But now I'm going to ask a nasty question: how do you know that two of these will never get themselves to be running at once? Like startViewChangeCosi blocks for x seconds, at time x sendNewView somehow gets called again, the second goroutine starts and then startViewChangeCosi goes fast this time, and now we have two createViewChangeBlocks racing?

Reply:
For that scenario to happen, the time it takes to do `sendNewView` must be longer than the time it takes for a successful view-change. Because, during a single view-change, every honest node will only send the new-view message once. For the race to happen, it'll have to be in two different view-changes. However, if it's in two different view-changes, then one view-change must have completed successfully. If the attacker doesn't have full control of the network delay (which is the assumption in the PBFT view-change, otherwise the exponential increase in timeout won't work), then the first  `sendNewView` must have completed, even if other nodes didn't accept it.